File: Data Structures & Algorithms/eating-bananas/submission-0.cpp
class Solution {
public:
    bool isValid(int mid, vector<int> &nums, int h){
    long long ans = 0;
    for(auto it: nums)
        ans += (it + mid -1)/mid;
    
    return ans <= h;
}

int minEatingSpeed(vector<int> nums, int h) {
    int l=1, r=*max_element(nums.begin(), nums.end());

    while(l < r){
        int mid = l + (r-l)/2;

        if(isValid(mid, nums, h))
            r=mid;
        else
            l=mid+1;
    }
    return l;
    }
};

File: Data Structures & Algorithms/combination-target-sum/submission-0.java
class Solution {

    public void solve(int index, int target, int[] nums, 
        List<Integer> tmp, List<List<Integer>> ans, int sum){

            if(index == nums.length){
                if(sum == target){
                    ans.add(new ArrayList<>(tmp));
                }
                return;
            }
            else if(target == sum){
                ans.add(new ArrayList<>(tmp));
                return ;
            }     
            else if(sum > target) return;

            tmp.add(nums[index]);
            solve(index, target, nums, tmp, ans, sum+nums[index]);
            tmp.remove(tmp.size()-1);
            solve(index+1, target, nums, tmp, ans, sum);
    } 

    public List<List<Integer>> combinationSum(int[] nums, int target) {
        List<Integer> tmp = new ArrayList<>();
        List<List<Integer>> ans = new ArrayList<>();

        solve(0, target, nums, tmp, ans, 0);
        return ans;
    }
}
